lib/propolis/src/enlightenment/bhyve.rs

```diff
@@ -0,0 +1,58 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Provides a bhyve-compatible guest-hypervisor interface.
+//!
+//! This interface supplies no special enlightenments; it merely identifies
+//! itself as a bhyve hypervisor in CPUID leaf 0x4000_0000.
+
+use cpuid_utils::{
+    bits::HYPERVISOR_BASE_LEAF, CpuidIdent, CpuidSet, CpuidValues,
+};
+
+use crate::{
+    common::{Lifecycle, VcpuId},
+    enlightenment::Enlightenment,
+    msr::{MsrId, RdmsrOutcome, WrmsrOutcome},
+};
+
+pub struct BhyveGuestInterface;
+
+impl Lifecycle for BhyveGuestInterface {
+    fn type_name(&self) -> &'static str {
+        "bhyve-guest-interface"
+    }
+}
+
+impl Enlightenment for BhyveGuestInterface {
+    fn add_cpuid(&self, cpuid: &mut CpuidSet) -> anyhow::Result<()> {
+        let old = cpuid
+            .insert(
+                CpuidIdent::leaf(HYPERVISOR_BASE_LEAF),
+                CpuidValues {
+                    eax: HYPERVISOR_BASE_LEAF,
+                    ebx: 0x76796862,
+                    ecx: 0x68622065,
+                    edx: 0x20657679,
+                },
+            )
+            .map_err(|_| {
+                anyhow::anyhow!("reserved leaf 0x4000_0000 already in map")
+            })?;
+
+        if old.is_some() {
+            anyhow::bail!("reserved leaf 0x4000_0000 already in map");
+        }
```

Comment:
I was a bit unsure why these both returned the same error, so I had to go re-read the `CpuidSet` implementation to figure this out. It's because `insert` returns an error if there's a conflict between the inserted values and whether or not an already-present value would have subleaves, and here, if we ever see that error, it implies that there's a value present, and we expect there not to be. So, using the same error message makes sense upon consideration, but it _might_ be nice to have a comment here writing that down for future readers?

Reply:
Fixed in 616ed70.